### ISSUE TITLE
Work plan: finetune π0.5 and GR00T N1.7 for the OpenArm embodiment on MI300X

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/dexter-labs#2]
Finetune π0.5 and GR00T N1.7 for the OpenArm bimanual embodiment on MI300X, and evaluate both **closed loop** in MuJoCo — not against recorded trajectories.

Neither checkpoint can drive OpenArm today. This is not a tuning problem, it is an architectural one, and the two models are blocked differently:

- **GR00T N1.7 is hard-blocked.** Its state encoder, action encoder and action decoder are `CategorySpecificLinear` — 32 *per-embodiment weight slices* indexed by `embodiment_id` (`state_encoder.layer1.W` is `[32, 132, 1024]`). `new_embodiment` is absent from the base checkpoint's `embodiment_id.json` entirely. There is no prompt that reaches those weights.
- **π0.5 is soft-blocked.** It has no per-embodiment tensors at all — one shared `action_in_proj [1024, 32]` over a padded 32-wide space — so OpenArm's 16 dims *will* run through it. But `pi05_droid` predicts **joint velocities**, which are kinematics-specific. It will produce confident nonsense. Finetune from `pi05_base`, not from the DROID checkpoint.

There is no in-context-learning escape: neither model has any mechanism to condition on demonstrations.

---

## Why MI300X changes the calculus

192 GB of HBM3 removes the memory constraint that forces LoRA elsewhere. Full finetunes fit on a **single** GPU.

| | trainable | optimizer state @ AdamW | fits 1×MI300X |
|---|---|---|---|
| GR00T, NVIDIA's recipe (backbone frozen) | 1.62 B | ~26 GB + 3 GB frozen | yes, with room for batch ≥128 |
| GR00T, everything unfrozen | 3.14 B | ~50 GB | yes |
| π0.5 full finetune | 3.62 B | ~58 GB | yes (openpi's own README says >70 GB, assuming an 80 GB A100/H100) |

An 8×MI300X node (1.5 TB) reproduces NVIDIA's published global batch of 640 without sharding tricks. **Do not start with LoRA.** It exists to work around memory we do not have a shortage of, and it costs quality.

NVIDIA's own published LIBERO recipe freezes the VLM, and it reaches 195/200. Start there:

```
tune_llm=False   tune_visual=False
tune_projector=True   tune_diffusion_model=True   tune_vlln=True
```

---

## Phase 0 — bring-up (blocking, ~1 day)

**T0.1 · ROCm + framework bring-up on MI300X (gfx942).**
Deliverable: a documented environment that runs GR00T N1.7 and π0.5 *inference* on one MI300X and reproduces a known score.
DoD: `nvidia/GR00T-N1.7-LIBERO` scores ≥95% on `libero_spatial` closed loop, matching what we already measured on gfx1151 (50/50). If it does not, stop — the environment is wrong, not the model.
Note: gfx942 is CDNA3, not the RDNA3.5 we developed on. Do not assume our gfx1151 workarounds apply; **re-measure, do not port assumptions.** In particular MIOpen's missing Conv3d solver and the power-of-two head-dim limit were gfx1151-specific.

**T0.2 · Decide the π0.5 training framework.** ⚠️ **Needs a human decision.**
openpi's trainer is **JAX**. The standing constraint on this project is PyTorch-only. Options:
(a) LeRobot's PyTorch π0.5 training path — consistent with the constraint, less battle-tested for training;
(b) openpi on JAX-ROCm — upstream's own recipe, but a second framework and an unproven ROCm path.
Do not start Phase 3b until this is answered.

**T0.3 · Stand up the OpenArm MuJoCo environment.**
`enactic/openarm_simulation` ships `openarm_mujoco/v1/openarm_bimanual.xml`.
DoD: an env that resets to a fixed initial state, steps a 16-dim action, renders the five cameras headless, and exposes a **scripted success predicate**. Without a success predicate there is no closed-loop metric and this whole plan degrades to open-loop scoring — see *Rules*.
Open question: does enactic ship a success/task definition, or must we write one? Resolve early; it is on the critical path.

---

## Phase 1 — data (the actual bottleneck)

**T1.1 · Assess the public dataset.**
`enactic/openarm-2-cell-pick_up_cube_mujoco-lerobot`: 30 episodes, 5405 frames @ 30 Hz, 16-dim state and action, five cameras (`ceiling`, `head_left`, `head_right`, `wrist_left`, `wrist_right`), 155 MB, and a `success` column.
DoD: a written verdict on whether 30 episodes of one task suffices. **Expected answer: no.** NVIDIA's LIBERO finetunes trained on the full demo suites; openpi's `pi05_libero` ran batch 256 × 30K steps. Budget ~100 demos per task.

**T1.2 · Collect or generate more demonstrations.**
Either teleoperate (LeRobot added native OpenArm leader/follower support in v0.5.0) or script demonstrations in MuJoCo.
DoD: ≥100 episodes per task across ≥3 tasks, in LeRobot v3 format, with a held-out split that is **held out by episode, never by frame**.

**T1.3 · Dataset integrity check.**
DoD: a report confirming every episode decodes, every camera stream covers every episode, and state/action ranges are sane per dimension.
Trap, from experience on DROID: a partially-downloaded slice indexed 1074 episodes while the downloaded videos covered 327, and seeking a missing episode returned *a real-looking frame from a different episode*. It cost a week of chasing a "model problem". Verify availability up front and refuse to sample what cannot be decoded.

---

## Phase 2 — the embodiment contract (highest bug density per line)

Every item in this phase is silent when wrong: the code runs and the policy is subtly worthless.

**T2.1 · Write `modality.json` for OpenArm.**
Maps the 16-dim state/action into named keys with index ranges, in the order the checkpoint concatenates them. Compare against LIBERO's, which splits `x,y,z,roll,pitch,yaw,gripper` where `gripper` spans `start:6,end:8` for state and `6:7` for action.
DoD: round-trip test — split a state vector into keys and reassemble it, and assert the result is bit-identical to the input.

⚠️ **`roll/pitch/yaw` in NVIDIA's schema is the *axis-angle*, not Euler angles.** Their own LIBERO wrapper fills those three from `quat2axisangle`. Reading the names literally and converting to Euler runs fine and is wrong.

**T2.2 · Choose and justify the camera set.**
OpenArm ships five streams; GR00T's shipped embodiments use ≤3 video keys. Picking is a modelling decision, not a detail — write down the reason.
Note: a longer image context is exactly where [the cross-attention bug we just fixed](https://github.com/raikonenfnu/dexter-labs/commit/1a98c95) bit. Five cameras would have made it far worse.

**T2.3 · Write the adapters.**
One per policy, following the existing `dexter/vla/groot/libero.py` and `dexter/vla/pi05_libero.py` pattern: cameras, `state_from`, action decode, and any gripper convention.
DoD: **a test asserting both adapters feed their models the same numbers**, as `test/test_libero.py::test_both_policies_are_fed_the_same_eight_numbers` does today. If the two disagree, one of them is feeding a distribution its model never saw.

⚠️ Gripper conventions differ per model *and* per simulator. GR00T's LIBERO path needs rescale `[0,1]→[-1,1]`, then binarise, then invert, because the RLDS loader aligned 0 with *close* while the controller reads −1 as *open*. π0.5 needs none of it. Determine OpenArm's convention empirically and test it.

⚠️ Image orientation. LIBERO needs a 180° rotation (`[::-1, ::-1]`): one flip undoes MuJoCo's bottom-up render, the other matches how the training data was converted. Establish OpenArm's convention by looking at the data, not by assuming.

**T2.4 · Compute normalisation statistics.**
GR00T uses q01/q99 percentiles, not min/max (`use_percentiles: true`). π0.5 needs openpi-format norm stats.
⚠️ Both `pi05_droid` and `pi05_libero` ship with **empty** `normalizer_processor.features`. Statistics do not travel with the checkpoint. Compute them from the training split and never from the eval split.

---

## Phase 3a — GR00T finetune (start here; it is the better-supported path)

`NEW_EMBODIMENT` is a first-class finetuning tag, the checkpoint already contains bimanual precedents (G1 and R1-Pro use left/right wrist eef + hands, structurally close to OpenArm), and the frozen-backbone recipe is NVIDIA's own.

```bash
bash examples/finetune.sh \
  --base-model-path nvidia/GR00T-N1.7-3B \
  --dataset-path <openarm-lerobot> \
  --modality-config-path <modality.json> \
  --embodiment-tag NEW_EMBODIMENT \
  --output-dir <out>
```

**T3a.1 · Overfit one batch.** Before any real run: train on a single batch until loss ≈ 0.
DoD: it overfits. If it cannot, the data pipeline or the contract is broken and a long run will only hide that.

**T3a.2 · Short pilot.** ~1K steps, batch 32, one GPU.
DoD: loss curve descends; a checkpoint loads; closed-loop success is **> 0%**. Not "good" — just non-zero, which separates "wired up" from "wired up wrong".

**T3a.3 · Full run.** Scale batch to the hardware (≥128 on one GPU, 640 across a node). NVIDIA used 20K steps for LIBERO.
DoD: closed-loop success on held-out initial states, reported with per-task breakdown.

---

## Phase 3b — π0.5 finetune (blocked on T0.2)

Finetune from **`pi05_base`**, not `pi05_droid`. Same overfit-one-batch → pilot → full-run ladder.

⚠️ `pi05_libero` ships **no token embedding**: PaliGemma ties it to the LM head and the checkpoint stores only the head, so LeRobot's loader warns and leaves 526M parameters randomly initialised. Verify before assuming — in `pi05_droid`, which stores both, the two tensors are bit-identical. Check whatever base checkpoint you load.

⚠️ Horizon comes from the training config, not the checkpoint metadata. `pi05_libero`'s config says `chunk_size: 50` while openpi trained it with `action_horizon=10`; the action tokens attend to each other, so asking for 50 perturbs the first 10 as well.

---

## Phase 4 — closed-loop evaluation

**T4.1 · Port the sim harness.** `dexter/sim/` already provides the policy↔simulator socket split, the rollout loop and per-task reporting. Add `dexter/sim/openarm.py` beside `libero.py`.
Keep the two-virtualenv split: simulator stacks pin old numpy/gym, policy stacks pin new torch. This is what openpi and NVIDIA both do.

**T4.2 · Score both policies, each at its own published cadence.** Forcing one re-planning cadence on both looks fairer and misrepresents both.

**T4.3 · Report.** Success rate with per-task breakdown, latency against the control-rate budget, and a rendered successful episode per policy.

---

## Rules for agents working this issue

1. **An open-loop metric is not evidence of correctness.** On DROID, a version of our GR00T port with a cross-attention bug that hid ~90% of the image scored **+0.944** correlation against the reference's +0.920 — it ranked *better* while badly broken. Fixed, it scores +0.920 and 50/50 in closed loop. Task success caught what correlation could not.
2. **Always print the trivial baseline.** Do-nothing scores surprisingly well on near-stationary teleop.
3. **Differential-test against the reference implementation, in fp32.** In bf16 the reference's own rounding (~4% by vision block 24) is larger than the class of bug worth finding.
4. **Module-level parity is necessary, not sufficient.** Our backbone matched at cos +0.99994 and the head at cos +0.99993 while the composition was 16 points worse end to end. Diff the whole pipeline too.
5. **Pin versions and record why.** mujoco must be 3.2.x for robosuite 1.4: newer pybind11 makes `jnt_type[i] in (mjJNT_HINGE, …)` return False for a numpy integer, and every environment dies on an assertion that reads like a malformed model.
6. **Verify weight loading.** Initialise to zero and assert nothing stayed zero. A loader that silently reshapes a mismatched tensor, or a `from_pretrained` that warns and continues, will produce a model that runs and is worthless.
7. **Never sample data you have not proven decodable.**
8. **Report failures with the output.** A run that did not converge is a result; a run that "looks fine" without a number is not.

---

## Open risks

- **Data volume is the schedule.** 30 public episodes will validate plumbing, not produce a policy. Everything after Phase 1 is gated on T1.2.
- **T0.2 is unresolved** and blocks the entire π0.5 track.
- **Success predicate.** If `openarm_simulation` ships no scripted success check, T0.3 grows considerably and Phase 4 cannot start.
- **gfx942 is not gfx1151.** Re-measure; do not port our workarounds on faith.
- Community GR00T finetunes on OpenArm already exist (`asterism45/gr00t_bi_openarm_*`). Unvetted, but worth an hour before training anything.

## References

- OpenArm: [docs](https://docs.openarm.dev/) · [hardware](https://github.com/enactic/openarm) · [simulation](https://github.com/enactic/openarm_simulation) · [LeRobot integration](https://huggingface.co/docs/lerobot/openarm)
- Dataset: `enactic/openarm-2-cell-pick_up_cube_mujoco-lerobot`
- Prior art in this repo: `dexter/sim/` (LIBERO closed loop), `dexter/vla/groot/libero.py`, `test/test_groot_parity.py`
